### PR TITLE
[master] visibility_control & portable installation.

### PR DESCRIPTION
**Pull request by seanyen**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1634  |
| Primary OS tested on | Windows |
| Robotic platform tested on | turtlebot on Gazebo Simulation |

---

## Description of contribution in a few bullet points

* Added visibility control for properly building shared library on Windows. https://index.ros.org/doc/ros2/Tutorials/Ament-CMake-Documentation/#building-libraries-on-windows
* Corrected the executable and libraries installation location for better portability support. https://index.ros.org/doc/ros2/Tutorials/Ament-CMake-Documentation/#building-a-library

* The bottom line is that this change should not break other platform builds.

## Description of documentation updates required from your changes

* None.

---

## Future work that may be required in bullet points

* More coming changes are staged at this branch: https://github.com/ms-iot/navigation2/tree/seanyen/windows Let me know how I can break them down to review-able changes.

### CODE REVIEW

**Comment by seanyen** [comment]
@SteveMacenski Let me know logistically what branch I should target and anything I should know. This change is probably the one touching the most of the sources for Windows port, so I want to tackle this first. :)

File: nav2_util/include/nav2_util/lifecycle_utils.hpp [hunk 1]
```diff
@@ -42,7 +43,7 @@ void startup_lifecycle_nodes(
 /**
  * \param[in] nodes A ':' seperated list of node names. eg. "/node1:/node2"
  */
-void startup_lifecycle_nodes(
+static void startup_lifecycle_nodes(
```
[the remark this answers is not in the document]
Comment (seanyen) [reply nav2_util/include/nav2_util/lifecycle_utils.hpp]: Sorry I forgot to remove this change. Removing it now.

**Comment by SteveMacenski** [comment]
@seanyen you should retarget this to master, we don't accept PRs directly to distribution branches. It'll trickle into Foxy and cherry picked to eloquent

File: nav2_dwb_controller/dwb_plugins/CMakeLists.txt [hunk 2]
```diff
@@ -63,7 +63,7 @@ endif()
 install(TARGETS simple_goal_checker stopped_goal_checker standard_traj_generator
         ARCHIVE DESTINATION lib
         LIBRARY DESTINATION lib
-        RUNTIME DESTINATION lib/${PROJECT_NAME}
+        RUNTIME DESTINATION bin
```
Comment (SteveMacenski) [inline nav2_dwb_controller/dwb_plugins/CMakeLists.txt:66]: libraries should go in bin?

**Comment by seanyen** [comment]
> What's this visibility control thing? How should I plan on dealing with this in the future?

Here is a comprehensive explanation for Linux developer :)
https://gcc.gnu.org/wiki/Visibility

From my perspective, the GCC and MSVC made a opposing default decision on how to decide what symbols to be visible of a shared library for the linker linkage. From the guide, for GCC, one can specify `-fvisibility=hidden` to the toolchain to enforce the similar policy that MSVC. Perhaps that is something we can do, so when someone adds an new interface, CI can catch the visibility control earlier.

File: nav2_dwb_controller/dwb_plugins/CMakeLists.txt (hunk 2 above)
Comment (seanyen) [reply nav2_dwb_controller/dwb_plugins/CMakeLists.txt:66]: Yes, for a shared library (`.DLL` files), in general it should go to the `bin` folder for the install space (which is specified in `RUNTIME DESTINATION`), and the corresponding imported library (`.LIB` files) goes to `lib` folder (which is specified in `ARCHIVE DESTINATION`). On Windows, the symbols information is stored in the imported lib and the MSVC linker is linking against it instead of the shared lib.

REF: https://cmake.org/cmake/help/latest/command/install.html

**Comment by seanyen** [comment]
@SteveMacenski Just want to quick check, I saw some test failures on and off from the CI for `The test did not generate a result file.` and are they flaky tests? Just want to make sure it is not a regression from this change.

**Comment by SteveMacenski** [comment]
@seanyen, yes, those are just 2 flaky tests, https://github.com/ros-planning/navigation2/issues/1585. I haven't had the cycles to fix them. No need to worry about it.

**Comment by SteveMacenski** [comment]
@seanyen just following up here

**Comment by seanyen** [comment]
@SteveMacenski Sorry I didn't get back to you earlier. I ack'ed your concern and I agreed that we should come up with some more sustainable way for the visibility control thing. I will find some time next week to iterate on what I can do by adding something in `nav2_common`.

**Comment by SteveMacenski** [comment]
Thanks. If that works out, it would be nice to actually just  have a generic tools available to generate this stuff, or have CI generate this stuff and include it in pushes to master. 

Since Windows is the only OS that seems to need this, its hard for me as a Linux user to really appreciate or debug this. There are (currently) so few windows power-users involved to really  help maintain to keep this stuff up and running.

**Comment by SteveMacenski** [comment]
Hi, just following up on this quickly a couple of weeks later.

**Comment by seanyen** [comment]
@SteveMacenski Sorry for the overdue on the response. I decided to start a new pull request here: #1704. In the new pull request, it is using CMake built-in features to take care of the visibility control and it should largely minimize the maintainer effort keeping the visibility up-to-dated, and in the new pull request, I include all the initial required Windows changes.

**Comment by SteveMacenski** [comment]
Amazing. Thanks for the attention- I’ll review tomorrow.
